bpf/grantedv2.h

```diff
@@ -0,0 +1,172 @@
+	 * flow entry comes from next_renewal_ms.
+	 */
+	uint32_t next_renewal_ms;
+	/*
+	 * How many milliseconds (unit) GK must wait
+	 * before sending the next capability renewal
+	 * request.
+	 */
+	uint32_t renewal_step_ms;
+	/*
+	 * If true, do not encapsulate granted packets and send them directly
+	 * to their destinations whenever it is possible.
+	 * When a capability is expiring, its granted packets must go
+	 * through Grantor servers.
+	 */
+	bool direct_if_possible;
+} __attribute__ ((packed));
+
+struct grantedv2_state {
+	/* When @budget_byte is reset. */
+	uint64_t budget_renew_at;
+	/*
+	 * When @budget1_byte is reset,
+	 * add @tx1_rate_kib_cycle * 1024 bytes to it.
+	 */
+	uint32_t tx1_rate_kib_cycle;
+	/*
+	 * When @budget2_byte is reset,
+	 * reset it to @tx2_rate_kib_cycle * 1024 bytes.
+	 */
+	uint32_t tx2_rate_kib_cycle;
+	/* How many bytes @src can still send in current cycle. */
+	int64_t budget1_byte;
+	/*
+	 * How many bytes @src can still send in current cycle in
+	 * the secondary channel.
+	 */
+	int64_t budget2_byte;
+	/*
+	 * When GK should send the next renewal to
+	 * the corresponding grantor.
+	 */
+	uint64_t send_next_renewal_at;
+	/*
+	 * How many cycles (unit) GK must wait before
+	 * sending the next capability renewal request.
+	 */
+	uint64_t renewal_step_cycle;
+	/*
+	 * If true, do not encapsulate granted packets and send them directly
+	 * to their destinations whenever it is possible.
+	 */
+	bool direct_if_possible;
+};
+
+static inline uint64_t
+grantedv2_init_inline(struct gk_bpf_init_ctx *ctx)
+{
+	struct gk_bpf_cookie *cookie = init_ctx_to_cookie(ctx);
+	struct grantedv2_params params = *(struct grantedv2_params *)cookie;
+	struct grantedv2_state *state = (struct grantedv2_state *)cookie;
+
+	RTE_BUILD_BUG_ON(sizeof(params) > sizeof(*cookie));
+	RTE_BUILD_BUG_ON(sizeof(*state) > sizeof(*cookie));
+
+	state->budget_renew_at = ctx->now + cycles_per_sec;
+	state->tx1_rate_kib_cycle = params.tx1_rate_kib_sec;
+	state->tx2_rate_kib_cycle = params.tx2_rate_kib_sec;
+	state->budget1_byte = (int64_t)params.tx1_rate_kib_sec * 1024;
+	state->budget2_byte = (int64_t)params.tx2_rate_kib_sec * 1024;
+	state->send_next_renewal_at = ctx->now +
+		params.next_renewal_ms * cycles_per_ms;
+	state->renewal_step_cycle = params.renewal_step_ms * cycles_per_ms;
+	state->direct_if_possible = params.direct_if_possible;
+
+	return GK_BPF_INIT_RET_OK;
+}
+
+static inline uint64_t
+grantedv2_pkt_begin(const struct gk_bpf_pkt_ctx *ctx,
+	struct grantedv2_state *state, uint32_t pkt_len)
+{
+	if (ctx->now >= state->budget_renew_at) {
+		int64_t max_budget1 = (int64_t)state->tx1_rate_kib_cycle * 1024;
+		int64_t cycles = ctx->now - state->budget_renew_at;
+		int64_t epochs = cycles / cycles_per_sec;
+
+		state->budget_renew_at = ctx->now + cycles_per_sec -
+		       (cycles % cycles_per_sec);
+		state->budget1_byte += max_budget1 * (epochs + 1);
+		if (state->budget1_byte > max_budget1)
+			state->budget1_byte = max_budget1;
```

Comment:
It's not really about precision, but about the fact that the primary limit may go negative. Consider the following case, the primary limit goes deep into a negative value. But, then, the source stops sending traffic and waits for the primary limit to become positive again. The code here is accounting for the fact that, during this waiting interval, there won't be an event to increment the primary limit, so the code has to find out how many epochs have passed and properly account for them. This extra code is not needed for the secondary limit because whenever an epoch passes, the secondary limit goes to its maximum value because the secondary limit can never go below zero.

Reply:
Ah, I see, now the concept of counting the epochs make sense. Thank you.